## What is Object Oriented Programming?

Object-Oriented Programming (OOP) is a way of writing programs that models the real world using **objects**.

Before OOP, code was written **procedurally** like one long list of instructions, with data passed from function to function. As programs grew larger, this became very hard to maintain. OOP lets us groups related data and functions together into a single unit called an **object**. This makes code easier to read, reuse, and extend.

### Procedural vs OOP : Side by Side

```python
# Procedural way
def deposit_money(account_name, balance, amount):
    balance += amount
    return balance

# OOP way: the action belongs to the object
account.deposit(500)
```

| Aspect | Procedural | Object-Oriented |
|--------|-----------|----------------|
| Focus | Functions | Objects and interactions |
| Data | Passed freely between functions | Bundled inside the object |
| Code reuse | Copy-paste | Inheritance |
| Real-world modeling | Awkward | Natural |

### The 4 Pillars of OOP

1. **Encapsulation** : Bundle data and methods together; protect internal details
2. **Inheritance** : Build new classes by extending existing ones
3. **Polymorphism** : Same method name, different behavior depending on the object
4. **Abstraction** : Expose only what's necessary; hide the complexity

### How Big Tech Uses OOP

- **Google** : Every search query, Gmail message, and YouTube video is an object with attributes and methods
- **Amazon** : Each product, order, and shipment is an object interacting with others
- **Uber** : A `Ride` object ties together a `Driver`, a `Rider`, a `Car`, and a `Route`
- **Netflix** : A base class powers `Movie`, `Series`, and `Documentary` objects that all stream the same way

We will use a **banking application** as our main example throughout this notebook.

### **1. Classes and Objects**

A **class** is a blueprint which defines what an entity looks like and what it can do.
An **object** is a specific instance created from that blueprint.

**Analogy:**
- The **blueprint of a house** = a class → describes rooms, doors, windows
- **Your actual house** = an object → built from the blueprint, with its own address and color

From one blueprint, you can build many houses. From one class, you can create many objects.

**Key points:**
1. The class defines the **structure** (what data it holds) and **behavior** (what it can do)
2. Each object has its **own copy** of the data with its own values
3. All objects of the same class share the **same methods**
4. No memory is used when you define a class instead memory is only used when objects are created

#### `__init__` : The Constructor

`__init__` is a special method that Python calls **automatically** when you create an object.
It sets up the object's initial data (attributes).

```python
acc1 = Account('Rishabh', 101, 25000)
# Python automatically calls Account.__init__(acc1, 'Rishabh', 101, 25000)
```

#### `self` : The Object Itself

`self` is how a method(method is a function within a class) refers to **the specific object it is being called on**.

When you write `acc1.deposit(500)`, Python internally translates this to:
`Account.deposit(acc1, 500)` : it passes the object as the first argument automatically.

Every method inside a class must have `self` as its **first parameter**.

In [184]:
class Account:

    def __init__(self, name, accountnumber, balance):  # constructor
        self.name = name           # attribute: belongs to this specific object
        self.acc_no = accountnumber
        self.bal = balance

In [188]:
# Creating three different objects from the same class
acc1 = Account('Rishabh', 101, 25000)
acc2 = Account('Priya', 102, 67548)
acc3 = Account('Roshan', 103, 76000)

print(acc1.bal, acc1.name, acc1.acc_no)
print(acc2.bal, acc2.name, acc2.acc_no)
print(acc3.bal, acc3.name, acc3.acc_no)

25000 Rishabh 101
67548 Priya 102
76000 Roshan 103


### Another Example: A Student Class

Classes can represent **any** real-world entity, not just bank accounts.
Here is a `Student` class to reinforce that a class is just a blueprint:

In [190]:
class Student:

    def __init__(self, name, roll_number, marks):
        self.name = name
        self.roll_no = roll_number
        self.marks = marks     # marks out of 100

    def is_passing(self):
        return self.marks >= 50

    def get_grade(self):
        if self.marks >= 90:
            return 'A+'
        elif self.marks >= 75:
            return 'A'
        elif self.marks >= 60:
            return 'B'
        elif self.marks >= 50:
            return 'C'
        else:
            return 'F'

In [192]:
# Three different students, three different objects
s1 = Student('Alice', 1, 92)
s2 = Student('Bob', 2, 47)
s3 = Student('Carol', 3, 68)

print(s1.name, '| Grade:', s1.get_grade(), '| Passing:', s1.is_passing())
print(s2.name, '| Grade:', s2.get_grade(), '| Passing:', s2.is_passing())
print(s3.name, '| Grade:', s3.get_grade(), '| Passing:', s3.is_passing())

Alice | Grade: A+ | Passing: True
Bob | Grade: F | Passing: False
Carol | Grade: B | Passing: True


#### `__str__` : Controlling What `print()` Shows

- When you `print(acc1)` without defining `__str__`, Python shows something like:
`<__main__.Account object at 0x0000013183...>`  which is not very helpful.

- `__str__` is a **special Python method** (called a dunder method because of the double underscores).
It lets you control what appears when you print an object.

- Python calls `__str__` automatically when you use `print()` or `str()` on your object.
You never call `obj.__str__()` directly instead just use `print(obj)`.

In [194]:
class Account:

    def __init__(self, name, accountnumber, balance):
        self.name = name
        self.acc_no = accountnumber
        self.bal = balance

    def __str__(self):
        # This runs automatically when you do print(acc1)
        return f'Account #{self.acc_no} belongs to {self.name} | Balance: ${self.bal}'

In [196]:
acc1 = Account('Rishabh', 101, 25000)
acc2 = Account('Priya', 102, 67548)
acc3 = Account('Roshan', 103, 76000)

# Now print() shows something meaningful
print(acc1)
print(acc2)
print(acc3)

Account #101 belongs to Rishabh | Balance: $25000
Account #102 belongs to Priya | Balance: $67548
Account #103 belongs to Roshan | Balance: $76000


### Big Tech Example: Amazon's Product Class

Amazon has over 350 million products in their catalog. Every single product is an object.
The `Product` class defines the shared structure like ASIN, title, price, rating, while each object holds the unique data for one product.

An **ASIN** (Amazon Standard Identification Number) is the unique ID Amazon gives every product.
The `add_review` method shows how Amazon recalculates the average rating each time a new review comes in.

In [218]:
class Product:

    def __init__(self, asin, title, price, seller, rating=0.0, review_count=0):
        self.asin = asin           # unique product ID (like B09G9HDGY2)
        self.title = title
        self.price = price
        self.seller = seller
        self.rating = rating
        self.review_count = review_count

    def __str__(self):
        return f'[{self.asin}] {self.title} | ${self.price} | {self.rating}/5 ({self.review_count} reviews)'

    def apply_discount(self, percent):
        # Used during Prime Day, Black Friday etc.
        self.price = round(self.price * (1 - percent / 100), 2)
        print(f'Discount applied! on {self.title} and now the New price: ${self.price}')

    def add_review(self, new_rating):
        # Recalculate running average when a new review comes in
        total_combined_rating = self.rating * self.review_count + new_rating
        self.review_count += 1
        self.rating = round(total_combined_rating / self.review_count, 1)

In [220]:
# Two different product objects
laptop = Product('B09G9HDGY2', 'Apple MacBook Pro 14-inch', 1999.99, 'Apple Inc.', 4.8, 2341)
headphones = Product('B07Q9MJKBV', 'Sony WH-1000XM4 Headphones', 279.99, 'Sony', 4.7, 18923)

print(laptop)
print(headphones)

[B09G9HDGY2] Apple MacBook Pro 14-inch | $1999.99 | 4.8/5 (2341 reviews)
[B07Q9MJKBV] Sony WH-1000XM4 Headphones | $279.99 | 4.7/5 (18923 reviews)


In [222]:
laptop.apply_discount(15)     # Prime Day 15% off

headphones.add_review(5)      # new 5-star review
print(headphones)             # rating should update

Discount applied! on Apple MacBook Pro 14-inch and now the New price: $1699.99
[B07Q9MJKBV] Sony WH-1000XM4 Headphones | $279.99 | 4.7/5 (18924 reviews)


---

Let me explain to you the logic behind the add_review() method used above:

```python
def add_review(self, new_rating):
    total_score = self.rating * self.review_count + new_rating
    self.review_count += 1
    self.rating = round(total_score / self.review_count, 1)
```

#### The Problem It Solves

Amazon stores **only two values** per product, `self.rating` (current average) and `self.review_count` (total number of reviews). When a **new review comes in**, the average must be **recalculated** without having access to every individual rating ever given.


**But to recalculate the average you don't have to access all the old rating because it will result in slow processing and take up extra memory, if we are accessing all the rating everytime. So instead we can use a mathematical trick explained below**

#### The Mathematical Trick


> If you know the **current average** and the **total number of reviews**, you can reverse-engineer the **total rating** at any point.

>                         total_rating = average × count

For Sony headphones:
- `self.rating = 4.7`
- `self.review_count = 18923`
- So total rating ever given combined would be = `4.7 × 18923 = 88938.1`


#### Line 1: total_combined_rating = self.rating * self.review_count + new_rating

```python
total_combined_rating = 4.7 * 18923 + 5
            = 88938.1 + 5
            = 88943.1
```

- `self.rating * self.review_count` → **reverse-engineers** the sum of all past ratings from the current average
- `+ new_rating` → **adds the new rating (5 stars)** on top of that historical total
- Now `total_combined_rating` represents the **grand total of all ratings ever given including the new one**


#### Line 2: self.review_count += 1

```python
self.review_count = 18923 + 1 = 18924
```

- Increments the count by 1 to account for the newly added review


#### Line 3: self.rating = round(total_combined_rating / self.review_count, 1)

```python
self.rating = round(88943.1 / 18924, 1)
            = round(4.7002..., 1)
            = 4.7
```

- Divides the new total score by the new review count to get the **updated average**
- `round(..., 1)` → rounds to **1 decimal place**, just like how Amazon displays ratings (`4.7`, `4.8`, etc.)
- The result is stored back into `self.rating`, **overwriting the old average**

---

### **2. Class Variables vs Instance Variables**

There are two kinds of attributes (data) in a class:

| Type | Defined where | Shared? | Example |
|------|--------------|---------|---------|
| **Instance Variable** | Inside `__init__` with `self.` | No, unique per object | customer name, balance |
| **Class Variable** | At class level, outside `__init__` | Yes, same for ALL objects | bank name, IFSC code, counter |

**Memory tip:**
- Class variables = a **shared whiteboard** in an office. Everyone (all objects) sees the same thing.
- Instance variables = each person's **private notebook**. Different for everyone.

Changing a class variable affects **every object** of that class at once.

In [225]:
class Account:
    counter = 1              # class variable: shared by all Account objects
    bank_name = 'PyBank'    # class variable: same for all accounts
    ifsc_code = 'PYBN0001234'  # class variable: same branch for all

    def __init__(self, name, balance=0):
        self.name = name              # instance variable: unique per object
        self.acc_no = Account.counter # use class variable to assign account number
        Account.counter += 1          # increment the shared counter
        self.bal = balance            # instance variable: unique per object

    def __str__(self):
        return f'[{Account.bank_name}] Account #{self.acc_no} | {self.name} | ${self.bal}'

In [227]:
acc1 = Account('Rishabh', 25000)
acc2 = Account('Priya', 67548)
acc3 = Account('Roshan', 76000)

print(acc1)
print(acc2)
print(acc3)
print()

print('Total accounts created so far:', Account.counter - 1)
print('Bank IFSC Code:', Account.ifsc_code)  # same for all

[PyBank] Account #1 | Rishabh | $25000
[PyBank] Account #2 | Priya | $67548
[PyBank] Account #3 | Roshan | $76000

Total accounts created so far: 3
Bank IFSC Code: PYBN0001234


### Big Tech Example: Uber's Ride Counter and Surge Pricing

Uber processes over 25 million rides every day. Two types of data live at different levels:

- **Class variables:** `total_rides` (global count) and `surge_multiplier` (applies to all rides during peak hours)
- **Instance variables:** rider name, driver name, destination, base fare which will be unique per ride

When Uber activates surge pricing on a Friday night, they change `surge_multiplier` once and it immediately affects every new ride object created after that point.

In [172]:
class Ride:
    total_rides = 0        # class variable: tracks all rides ever created
    surge_multiplier = 1.0 # class variable: changed globally during peak hours

    def __init__(self, rider, driver, pickup, destination, base_fare):
        self.rider = rider            # instance variable
        self.driver = driver          # instance variable
        self.pickup = pickup          # instance variable
        self.destination = destination
        self.base_fare = base_fare
        Ride.total_rides += 1
        self.ride_id = Ride.total_rides

    def get_final_fare(self):
        # Multiply the base fare by the current surge multiplier
        return round(self.base_fare * Ride.surge_multiplier, 2)

    def __str__(self):
        return (f'Ride #{self.ride_id}: {self.rider} to {self.destination} '
                f'with {self.driver} | Fare: ${self.get_final_fare()}')

In [176]:
# Normal pricing
ride1 = Ride('Alice', 'Carlos', 'Marina District', 'SFO Airport', 45.00)
ride2 = Ride('Bob', 'Priya', 'Downtown', 'Oakland', 22.50)
print(ride1)
print(ride2)

Ride #4: Alice to SFO Airport with Carlos | Fare: $103.5
Ride #5: Bob to Oakland with Priya | Fare: $51.75


In [180]:
# Surge pricing kicks in (say, Friday night) and just one change affects all future rides
Ride.surge_multiplier = 2.3 #updating the class variable
print(f'\nSurge pricing activated: x{Ride.surge_multiplier}')

ride3 = Ride('Carol', 'Sam', 'Tenderloin', 'Caltrain', 15.00)
print(ride3)

print(f'Total rides so far: {Ride.total_rides}')


Surge pricing activated: x2.3
Ride #7: Carol to Caltrain with Sam | Fare: $34.5
Total rides so far: 7


## 3. Methods : Functions Inside a Class

Methods are just **functions defined inside a class**. They always receive `self` as the first argument so they can access and modify the object's data.

Methods let objects **do things**: deposit money, send a message, play a song.

In [229]:
class Account:

    def __init__(self, name, accountnumber, balance):
        self.name = name
        self.acc_no = accountnumber
        self.bal = balance

    def __str__(self):
        return f'Account #{self.acc_no} belongs to {self.name} & the account Balance: ${self.bal}'

    def deposit(self, amount):
        if amount > 0:
            self.bal += amount
            print(f"Deposited ${amount} to {self.name}'s account. New balance: ${self.bal}")
        else:
            print('Deposit amount must be positive.')

    def withdraw(self, amount):
        if amount <= 0:
            print('Withdraw amount must be positive.')
        elif amount > self.bal:
            print('Insufficient funds!')
        else:
            self.bal -= amount
            print(f"Withdrew ${amount} from {self.name}'s account. New balance: ${self.bal}")

    def transfer(self, amount, other_account):
        # Transfer from self to other_account
        if amount <= 0:
            print('Transfer amount must be positive.')
        elif amount > self.bal:
            print('Insufficient funds to transfer!')
        else:
            print(f'Transferring ${amount} from {self.name} to {other_account.name}...')
            self.withdraw(amount)
            other_account.deposit(amount)
            print(f'Transfer complete.')

In [231]:
acc1 = Account('Rishabh', 101, 25000)
acc2 = Account('Priya', 102, 67548)

print(acc1)
print(acc2)
acc1.deposit(10000)

Account #101 belongs to Rishabh & the account Balance: $25000
Account #102 belongs to Priya & the account Balance: $67548
Deposited $10000 to Rishabh's account. New balance: $35000


In [233]:
acc1.transfer(20000, acc2)
print()
print(acc1)
print(acc2)

Transferring $20000 from Rishabh to Priya...
Withdrew $20000 from Rishabh's account. New balance: $15000
Deposited $20000 to Priya's account. New balance: $87548
Transfer complete.

Account #101 belongs to Rishabh & the account Balance: $15000
Account #102 belongs to Priya & the account Balance: $87548


### **4. Encapsulation**

Encapsulation means **bundling data and methods together, and controlling who can access the data**.

#### Why Do We Need It?

Without encapsulation, anyone can change an object's data directly and even put invalid values:

In [34]:
# Without encapsulation
class Account:
    def __init__(self, name, balance):
        self.name = name
        self.balance = balance  # public attribute which anyone can change it

acc = Account('Alice', 5000)

# Anyone can corrupt the data since there is no validation at all
acc.balance = -99999999   # This should never be allowed
print('Balance after corruption:', acc.balance)  # -99999999

Balance after corruption: -99999999


Encapsulation solves this by making attributes **private** and forcing access through controlled methods.

### Access Levels in Python

| Convention | Syntax | Meaning |
|-----------|--------|---------|
| Public | `self.name` | Accessible from anywhere (default) |
| Protected | `self._name` | Single underscore = "please don't touch this" (convention only, not enforced) |
| Private | `self.__name` | Double underscore = name-mangled, harder to access from outside |

**Important note on private attributes:**
Python does not truly block access instead it uses **name mangling**.
`self.__balance` internally becomes `self._Account__balance`.
So you *can* still access it from outside, but you have to go out of your way and put extra intentional effort to do so.

### Getters and Setters

Since private attributes cannot be accessed directly, we provide controlled access through:
- **Getter method:** a method that *returns/shows* the private value and it could be written like (`get_balance()`)
- **Setter method:** a method that *sets* the private value after validating if the value to be set is valid (`set_balance()`)

In [237]:
class Account:

    def __init__(self, name, balance=0):
        self.name = name
        self.__balance = balance   # private attribute (double underscore)

    def deposit(self, amount):
        if amount > 0:
            self.__balance += amount
            print(f'${amount} deposited. New balance: ${self.__balance}')
        else:
            print('Deposit amount must be positive.')

    def withdraw(self, amount):
        if amount <= 0:
            print('Withdraw amount must be positive.')
        elif amount > self.__balance:
            print('Insufficient funds.')
        else:
            self.__balance -= amount
            print(f'${amount} withdrawn. New balance: ${self.__balance}')

    def get_balance(self):          # getter to help read the private value
        return self.__balance

    def set_balance(self, amount):  # setter to help write with validation
        if isinstance(amount, (int, float)) and amount >= 0:
            self.__balance = amount
            print(f'Balance updated to ${self.__balance}')
        else:
            print('Invalid amount. Balance must be a non-negative number.')

In [239]:
account = Account('John Doe', 100)
account.deposit(50)
account.withdraw(30)
print(f'Current balance: ${account.get_balance()}')

$50 deposited. New balance: $150
$30 withdrawn. New balance: $120
Current balance: $120



In [243]:
# Direct access to private attribute will fail
print(account.__balance)

AttributeError: 'Account' object has no attribute '__balance'

In [247]:
# Name mangling : Python renames "account.__balance" internally to _Account__balance
print(f'Try to access private variable via name mangling: ${account._Account__balance}')
print()

# Setter with validation
account.set_balance('abc')     # Invalid because string not allowed
account.set_balance(-500)      # Invalid since negative not allowed
account.set_balance(9999)      # Valid

Try to access private variable via name mangling: $9999

Invalid amount. Balance must be a non-negative number.
Invalid amount. Balance must be a non-negative number.
Balance updated to $9999


### Big Tech Example : Apple's Credit Card (Encapsulated Financial Data)

When you add a card to Apple Wallet, Apple never stores your full card number. The real card number is **private**. The app only shows the last 4 digits.

Stripe uses the same pattern: card numbers are tokenized and private; the public API only exposes the last 4 digits and expiration date.

Notice how we use getter methods (`get_masked_number()`, `get_available_credit()`, `get_utilization()`) to safely expose calculated information without revealing the private data.

In [253]:
class CreditCard:

    def __init__(self, holder_name, card_number, cvv, credit_limit):
        self.holder_name = holder_name
        self.__card_number = str(card_number)  # private since full number never to be exposed
        self.__cvv = cvv                       # private, not to be shown
        self.__balance = 0                     # outstanding amount you currently owe to the credit card company right now
        self.__credit_limit = credit_limit

    # Getters for controlled read-only access to calculated values
    def get_masked_number(self):
        # Only show last 4 digits like Apple Wallet
        return f'**** **** **** {self.__card_number[-4:]}'

    def get_available_credit(self):
        return self.__credit_limit - self.__balance

    def get_utilization(self):
        # Credit utilization percentage, important for calculating credit scores
        return round((self.__balance / self.__credit_limit) * 100, 1)

    def charge(self, amount, merchant):
        if amount <= 0:
            print('Charge amount must be positive.')
            return False #immediately exits the method and Python doesn't execute any line below it and hence __balance is untouched
        if amount > self.get_available_credit():
            print(f'Declined at {merchant}: need ${amount}, only ${self.get_available_credit()} available')
            return False
        self.__balance += amount
        print(f'Approved: ${amount} at {merchant} | Remaining credit: ${self.get_available_credit()}')
        return True #tells the calling function: "Charge was approved and processed successfully"

    def pay_bill(self, amount):
        # Pay as much as is owed (can't overpay)
        payment = min(amount, self.__balance)
        self.__balance -= payment
        print(f'Payment of ${payment} received. Remaining balance: ${self.__balance}')

    def __str__(self):
        return f'Card: {self.get_masked_number()} | {self.holder_name} | Utilization: {self.get_utilization()}%'

In [255]:
card = CreditCard('John Doe', '4111111111111234', 123, 5000)

card.charge(1500, 'Amazon')
card.charge(2000, 'Apple Store')
card.charge(2000, 'Best Buy')   # should be declined

Approved: $1500 at Amazon | Remaining credit: $3500
Approved: $2000 at Apple Store | Remaining credit: $1500
Declined at Best Buy: need $2000, only $1500 available


False

In [257]:
print(card)
print()

card.pay_bill(1000)
print(f'Utilization after payment: {card.get_utilization()}%')

Card: **** **** **** 1234 | John Doe | Utilization: 70.0%

Payment of $1000 received. Remaining balance: $2500
Utilization after payment: 50.0%


### **5. Inheritance**

Inheritance lets a new class (**child / subclass**) reuse and extend the code of an existing class (**parent / superclass**).

This models the **"is-a" relationship**: a `SavingsAccount` *is an* `Account`.

**Why it matters:**
Without inheritance, you would copy-paste `deposit()` and `withdraw()` into every account type.
With inheritance, you write it once in the parent and every child gets it automatically.

### `super()` : Calling and accessing the Parent's Method

When a child class has its own `__init__`, it usually needs to call the parent's `__init__` too so that the parent's attributes are properly accessible. `super()` does exactly this:

```python
class CurrentAccount(Account):
    def __init__(self, name, acc_no, balance, gender):
        super().__init__(name, acc_no, balance)  # run Account's __init__ first
        self.gender = gender                      # then add CurrentAccount's own data
```

### The 5 Types of Inheritance in Python

1. **Single** : One child inherits from one parent
2. **Multiple** : One child inherits from two or more parents
3. **Multilevel** : A chain: Grandparent → Parent → Child
4. **Hierarchical** : Multiple children inherit from the same parent
5. **Hybrid** : A mix of the above

#### **5.1 Single Inheritance**

In [66]:
class Account:

    def __init__(self, name, accountnumber, balance):
        self.name = name
        self.acc_no = accountnumber
        self.bal = balance
        self.numtrans = 0    # number of transactions done today
        self.maxtrans = 2    # max transactions allowed per day

    def deposit(self, amount):
        if amount > 0 and self.numtrans < self.maxtrans:
            self.bal += amount
            self.numtrans += 1
            print(f'Deposited ${amount}. Balance: ${self.bal}')
        elif self.numtrans >= self.maxtrans:
            print('Transaction limit reached for today.')
        else:
            print('Deposit amount must be positive.')

    def withdraw(self, amount):
        if self.numtrans >= self.maxtrans:
            print('Transaction limit reached for today.')
        elif amount <= 0:
            print('Withdraw amount must be positive.')
        elif amount > self.bal:
            print('Insufficient Funds!')
        else:
            self.bal -= amount
            self.numtrans += 1
            print(f'Withdrew ${amount}. Balance: ${self.bal}')

    def __str__(self):
        return f'Account #{self.acc_no} | {self.name} | ${self.bal}'


# SavingsAccount is also an Account and it inherits everything from Account
class SavingsAccount(Account):
    pass   # 'pass' means: no extra code needed, it uses all of Account class's methods as it is directly

# CurrentAccount is also an Account but with different limits
class CurrentAccount(Account):
    def __init__(self, name, accountnum, balance, gender):
        super().__init__(name, accountnum, balance)  # access and use Account's __init__ along with CurrentAccount's __init__
        self.maxtrans = 5   # override: current accounts allow more transactions
        self.gender = gender

In [68]:
sav1 = SavingsAccount('Vinay', 301, 90800)
cur1 = CurrentAccount('Meera', 401, 200000, 'Female')

print('Savings Account (max 2 transactions allowed)')
sav1.deposit(5000)   # transaction 1
sav1.deposit(5000)   # transaction 2
sav1.deposit(5000)   # refused since transaction limit reached

print()
print('Current Account (max 5 transactions allowed)')
cur1.deposit(10000)  # transaction 1
cur1.deposit(10000)  # transaction 2
cur1.deposit(10000)  # transaction 3

Savings Account (max 2 transactions allowed)
Deposited $5000. Balance: $95800
Deposited $5000. Balance: $100800
Transaction limit reached for today.

Current Account (max 5 transactions allowed)
Deposited $10000. Balance: $210000
Deposited $10000. Balance: $220000
Deposited $10000. Balance: $230000


#### **5.2 Multiple Inheritance**

In [264]:
# Real-world example: A Tesla is both a Vehicle and uses GPS and is Electric vehicle

class Vehicle:
    def __init__(self, make, model):
        self.make = make
        self.model = model

    def start(self):
        return f'{self.make} {self.model} is starting...'

class ElectricVehicle:
    def __init__(self, battery_kwh):
        self.battery = battery_kwh
        self.charge_level = 100   # percentage

    def get_range(self):
        # rough estimate: 2.5 miles per kWh at 100% charge
        miles = int(self.battery * (self.charge_level / 100) * 2.5)
        return f'{miles} miles'

    def recharge(self):
        self.charge_level = 100
        print(f'Battery fully recharged ({self.battery} kWh).')

class GPSNavigator:
    def get_location(self):
        return '37.7749 N, 122.4194 W'  # San Francisco

    def navigate(self, destination):
        return f'Routing to {destination} from {self.get_location()}'

# TeslaModel3 inherits from all three classes above showing multiple inheritance
class TeslaModel3(Vehicle, ElectricVehicle, GPSNavigator):
    def __init__(self, model, autopilot=True):
        Vehicle.__init__(self, 'Tesla', model)
        ElectricVehicle.__init__(self, battery_kwh=75)
        self.autopilot = autopilot

    def self_drive(self, destination):
        if self.autopilot:
            return f'Autopilot engaged. {self.navigate(destination)} | Range: {self.get_range()}'
        return 'Autopilot not available.'

tesla = TeslaModel3('Model 3')
print(tesla.start())                      # from Vehicle
print(tesla.get_range())                  # from ElectricVehicle
print(tesla.navigate('Palo Alto'))        # from GPSNavigator
print(tesla.self_drive('Palo Alto'))      # TeslaModel3's own method

Tesla Model 3 is starting...
187 miles
Routing to Palo Alto from 37.7749 N, 122.4194 W
Autopilot engaged. Routing to Palo Alto from 37.7749 N, 122.4194 W | Range: 187 miles


#### **5.3: Multilevel Inheritance** : Grandparent → Parent → Child. Each level adds more specific behavior

In [267]:
class Animal:
    def breathe(self):
        return 'Breathing oxygen.'

    def eat(self):
        return 'Eating food.'


class Mammal(Animal):           # Mammal is an Animal
    def feed_young_with_milk(self):
        return 'Feeding young with milk.'

    def regulate_temperature(self):
        return 'Maintaining body temperature.'


class Dog(Mammal):              # Dog is a Mammal (which is also an Animal)
    def bark(self):
        return 'Woof!'

    def fetch(self):
        return 'Fetching the ball!'


dog = Dog()
print(dog.breathe())               # from Animal (grandparent)
print(dog.feed_young_with_milk())  # from Mammal (parent)
print(dog.bark())                  # Dog's own method
print(dog.fetch())                 # Dog's own method

# isinstance() checks the full chain
print()
print('Is dog a Dog?', isinstance(dog, Dog))       # True
print('Is dog a Mammal?', isinstance(dog, Mammal)) # True 
print('Is dog an Animal?', isinstance(dog, Animal))# True

Breathing oxygen.
Feeding young with milk.
Woof!
Fetching the ball!

Is dog a Dog? True
Is dog a Mammal? True
Is dog an Animal? True


#### **5.4 Hierarchical Inheritance**: Multiple children inherit from the same parent. A bank has many account types, all sharing the same base Account

In [270]:
class BankAccount:
    def __init__(self, holder, acc_no, balance):
        self.holder = holder
        self.acc_no = acc_no
        self.balance = balance

    def __str__(self):
        # type(self).__name__ gives us 'SavingsAccount', 'CurrentAccount' etc.
        return f'{type(self).__name__} | {self.holder} | ${self.balance}'


class SavingsAccount(BankAccount):
    interest_rate = 0.04   # 4% annual interest

    def apply_interest(self):
        earned = self.balance * self.interest_rate
        self.balance += earned
        print(f'Interest earned: ${round(earned, 2)}. New balance: ${round(self.balance, 2)}')


class CurrentAccount(BankAccount):
    overdraft_limit = 10000   # allowed to go into debt up to $10,000

    def withdraw(self, amount):
        if amount <= self.balance + self.overdraft_limit:
            self.balance -= amount
            print(f'Withdrew ${amount}. Balance: ${self.balance}')
        else:
            print('Amount exceeds overdraft limit.')

class FixedDepositAccount(BankAccount):
    def __init__(self, holder, acc_no, balance, lock_months):
        super().__init__(holder, acc_no, balance)
        self.lock_months = lock_months

    def is_locked(self, months_passed):
        locked = months_passed < self.lock_months
        status = 'LOCKED' if locked else 'FD MATURED, can withdraw now'
        print(f'FD after {months_passed} months: {status}')
        return locked

In [272]:
savings = SavingsAccount('Alice', 101, 50000)
current = CurrentAccount('Bob', 102, 2000)
fd = FixedDepositAccount('Carol', 103, 100000, lock_months=12)

print(savings)
print(current)
print(fd)
print()

savings.apply_interest()
current.withdraw(10000)   # within overdraft limit
current.withdraw(5000)    # exceeds limit
fd.is_locked(6)
fd.is_locked(13)

SavingsAccount | Alice | $50000
CurrentAccount | Bob | $2000
FixedDepositAccount | Carol | $100000

Interest earned: $2000.0. New balance: $52000.0
Withdrew $10000. Balance: $-8000
Amount exceeds overdraft limit.
FD after 6 months: LOCKED
FD after 13 months: FD MATURED, can withdraw now


False

### Method Resolution Order (MRO)

When a class inherits from multiple parents that share a common ancestor,
Python needs a rule for which parent's method to call.
It uses the **MRO (Method Resolution Order)** where it essentially: checks left parent before right parent.

This is called the **Diamond Problem**:
```
        A
       / \
      B   C
       \ /
        D
```
If both B and C have `hello()`, and D inherits from both, which one does D use?
Python answers this with MRO: check D first, then B, then C, then A.

#### **5.5: Hybrid Inheritance + MRO**

In [77]:
class A:
    def hello(self):
        return 'Hello from A'

class B(A):
    def hello(self):
        return 'Hello from B'

class C(A):
    def hello(self):
        return 'Hello from C'

class D(B, C):  # inherits from B first, then C, here MRO determines which hello() wins
    pass

d = D()
print(d.hello())   # MRO says: check D → B → C → A. B is first to have hello(), so 'Hello from B'

# See the full resolution order
print()
print('MRO for D:')
for cls in D.__mro__:
    print(f'  -> {cls.__name__}')

Hello from B

MRO for D:
  -> D
  -> B
  -> C
  -> A
  -> object


## **6. Polymorphism**

**Polymorphism** means "many forms" where the same method name produces different behavior depending on the object it is called on.

#### Simple Example First

Before we look at banking, here is the clearest possible example of polymorphism. Three completely different animals all have a `speak()` method, but each one behaves differently:

In [89]:
class Cat:
    def speak(self):
        return 'Meow!'

class Dog:
    def speak(self):
        return 'Woof!'

class Duck:
    def speak(self):
        return 'Quack!'

# All three have same method name "speak()" but different behavior
animals = [Cat(), Dog(), Duck()]

for animal in animals:
    # We call the same method on each object...
    # but the output is different each time and that is polymorphism
    print(f'{type(animal).__name__}: {animal.speak()}')

Cat: Meow!
Dog: Woof!
Duck: Quack!


### Method Overriding : Polymorphism Through Inheritance

The most common form of polymorphism in OOP is **method overriding**: a child class redefines a method from its parent class.

In the example below, `SavingsAccount` and `CurrentAccount` both inherit `withdraw()` from `Account`, but each has its own version of it. When you call `withdraw()`, Python runs the **right version** based on the object's type.

In [275]:
class Account:
    def __init__(self, name, accountnumber, balance):
        self.name = name
        self.acc_no = accountnumber
        self.bal = balance
        self.numtrans = 0
        self.maxtrans = 2

    def deposit(self, amount):
        if amount > 0 and self.numtrans < self.maxtrans:
            self.bal += amount
            self.numtrans += 1
            print(f'[{type(self).__name__}] Deposited ${amount}. Balance: ${self.bal}')
        else:
            print('Deposit refused.')

    def withdraw(self, amount):
        # Default withdraw: just check funds and transaction limit
        if self.numtrans >= self.maxtrans:
            print('Transaction limit reached.')
        elif amount <= 0:
            print('Withdraw amount must be positive.')
        elif amount > self.bal:
            print('Insufficient Funds!')
        else:
            self.bal -= amount
            self.numtrans += 1
            print(f'[{type(self).__name__}] Withdrew ${amount}. Balance: ${self.bal}')

    def __str__(self):
        return f'{type(self).__name__} #{self.acc_no} | {self.name} | ${self.bal}'

class SavingsAccount(Account):
    # Override withdraw: savings accounts charge a $1 fee per withdrawal
    def withdraw(self, amount):
        fee = 1
        total = amount + fee
        if self.numtrans >= self.maxtrans:
            print('Transaction limit reached.')
        elif total > self.bal:
            print('Insufficient funds after applying $1 withdrawal fee.')
        else:
            self.bal -= total
            self.numtrans += 1
            print(f'[SavingsAccount] Withdrew ${amount} + ${fee} fee. Balance: ${self.bal}')

class CurrentAccount(Account):
    def __init__(self, name, accountnum, balance, gender):
        super().__init__(name, accountnum, balance)
        self.maxtrans = 5   # current accounts get more transactions
        self.gender = gender

    # Override withdraw: no fee, higher transaction limit
    def withdraw(self, amount):
        if self.numtrans >= self.maxtrans:
            print('Transaction limit reached.')
        elif amount <= 0:
            print('Amount must be positive.')
        elif amount > self.bal:
            print('Insufficient funds.')
        else:
            self.bal -= amount
            self.numtrans += 1
            print(f'[CurrentAccount] Withdrew ${amount}. Balance: ${self.bal}')

In [277]:
savings = SavingsAccount('Alice', 'SA123', 100)
current = CurrentAccount('Bob', 'CA456', 200, 'Male')

# Same method call - different behavior because of the object's type
savings.withdraw(20)    # charges $1 fee
current.withdraw(20)    # no fee

[SavingsAccount] Withdrew $20 + $1 fee. Balance: $79
[CurrentAccount] Withdrew $20. Balance: $180


### Duck Typing : Python's Most Flexible Form of Polymorphism

In Python, you don't even need a shared parent class for polymorphism.

If two objects both have the same method name, you can use them interchangeably.
Python only checks **at runtime** whether the method exists and it doesn't care about the type.

This is called **duck typing**: "If it walks like a duck and quacks like a duck, it's a duck."

**Big tech example:**
An e-commerce checkout needs to work with Stripe, PayPal, Apple Pay, and Crypto. None of these share a parent class but they all have a `.process()` method. That's all Python needs.

In [100]:
class StripePayment:
    def process(self, amount):
        return f'Stripe: processed ${amount} via card'

class PayPalPayment:
    def process(self, amount):
        return f'PayPal: processed ${amount} via PayPal account'

class ApplePayPayment:
    def process(self, amount):
        return f'Apple Pay: processed ${amount} via device token'

class CryptoPayment:
    def process(self, amount):
        # Convert dollars to BTC at a rough rate
        btc = round(amount / 60000, 6)
        return f'Crypto: processed {btc} BTC (≈${amount})'

def checkout(payment_method, amount):
    # Works with ANY object that has a .process() method
    # No isinstance() check needed
    result = payment_method.process(amount)
    print(f'  Payment confirmed: {result}')

print('Customer checkout ($99.99):')
for method in [StripePayment(), PayPalPayment(), ApplePayPayment(), CryptoPayment()]:
    checkout(method, 99.99)

Customer checkout ($99.99):
  Payment confirmed: Stripe: processed $99.99 via card
  Payment confirmed: PayPal: processed $99.99 via PayPal account
  Payment confirmed: Apple Pay: processed $99.99 via device token
  Payment confirmed: Crypto: processed 0.001667 BTC (≈$99.99)


### **7. Method Overriding, Method Overloading, and Operator Overloading**

These three concepts are related to polymorphism:

1. **Method Overriding** : A child class redefines a parent's method (we just saw this above)
2. **Method Overloading** : Same method name, different number of parameters (Python handles this differently)
3. **Operator Overloading** : Redefining what operators like `+`, `==`, `<` do for your objects

### Method Overloading

In languages like Java and C++, you can define two methods with the same name but different parameters, and the language picks the right one automatically.

Python does **not** support this directly. If you define two methods with the same name, the second one silently replaces the first.

We can simulate overloading using **default parameter values** or `*args`.

In [107]:
# The wrong way where Python replaces the first add() with the second
class Calculator:
    def add(self, a, b):
        return a + b

    def add(self, a, b, c):   # this replaces the add(a, b) above
        return a + b + c

calc = Calculator()
try:
    calc.add(3, 4)    # error: Python only knows the 3-argument version now
except TypeError as e:
    print(f'Error: {e}')

Error: Calculator.add() missing 1 required positional argument: 'c'


In [109]:
# Fix 1: Default parameter values
class Calculator:
    def add(self, a, b=0, c=0):
        return a + b + c

calc = Calculator()
print(calc.add(5))          # 5  (b and c default to 0)
print(calc.add(5, 10))      # 15 (c defaults to 0)
print(calc.add(5, 10, 15))  # 30

print()

# Fix 2: *args accepts any number of arguments (the most flexible approach)
class Calculator:
    def add(self, *args):
        # args is a tuple of all arguments passed in
        return sum(args)

calc = Calculator()
print(calc.add(5))               # 5
print(calc.add(5, 10))           # 15
print(calc.add(5, 10, 15, 20))   # 50

5
15
30

5
15
50


### Operator Overloading

Python's built-in operators like `+`, `==`, `<` already work differently depending on the type:

In [111]:
print('Python' + 'Programming')    # string: concatenation
print([1, 2, 3] + [4, 5, 6])      # list: joins two lists
print(10 + 12)                      # numbers: addition

# Same operator (+), three completely different behaviors and that is what operator overloading is

PythonProgramming
[1, 2, 3, 4, 5, 6]
22


### **8. Abstraction**

Abstraction means **hiding the complexity and showing only what is necessary**.

When you press the brake pedal in a car, you don't need to know about the hydraulic system, brake pads, or ABS sensors, you just know pressing it slows the car. That is abstraction.

### Abstraction vs Encapsulation : Common Confusion

These two are often confused with each other. Here is the key difference:

| | Abstraction | Encapsulation |
|--|------------|---------------|
| What it hides | **Implementation** (how things work) | **Data** (internal state) |
| Answers | "What does this do?" | "Who can access this data?" |
| Example | Brake pedal hides hydraulics | `__balance` hides the bank balance |

### How We Implement Abstraction in Python

We create a **base class** that:
1. Defines which methods MUST exist (the contract)
2. Leaves those methods intentionally unimplemented
3. If a subclass forgets to implement a required method, Python raises an error

We signal an unimplemented method by using `raise NotImplementedError(...)`.

**Note:** Python has a module called `abc` (Abstract Base Classes) that can enforce this more strictly.
We will cover that later when we learn about decorators. For now, `NotImplementedError` achieves the same idea.

In [119]:
class Account:
    # This is a 'template' base class. It defines WHAT every account must do
    # but leaves the HOW to each subclass.

    def __init__(self, name, accountnumber, balance):
        self.name = name
        self.acc_no = accountnumber
        self.bal = balance

    def deposit(self, amount):
        # Not implemented here as every subclass MUST implement this
        raise NotImplementedError('Subclasses must implement their own deposit() method')

    def withdraw(self, amount):
        raise NotImplementedError('Subclasses must implement their own withdraw() method')

    def __str__(self):
        return f'Account #{self.acc_no} | {self.name} | ${self.bal}'


class SavingsAccount(Account):
    # SavingsAccount implements the required methods its own way
    def deposit(self, amount):
        if amount > 0:
            self.bal += amount
            print(f'[Savings] Deposited ${amount}. Balance: ${self.bal}')
        else:
            print('Deposit amount must be positive.')

    def withdraw(self, amount):
        fee = 1
        total = amount + fee
        if amount > 0 and total <= self.bal:
            self.bal -= total
            print(f'[Savings] Withdrew ${amount} + ${fee} fee. Balance: ${self.bal}')
        else:
            print('[Savings] Insufficient funds or invalid amount.')


class CurrentAccount(Account):
    # CurrentAccount implements the same methods but differently (no fee)
    def deposit(self, amount):
        if amount > 0:
            self.bal += amount
            print(f'[Current] Deposited ${amount}. Balance: ${self.bal}')
        else:
            print('Deposit amount must be positive.')

    def withdraw(self, amount):
        if amount > 0 and amount <= self.bal:
            self.bal -= amount
            print(f'[Current] Withdrew ${amount}. Balance: ${self.bal}')
        else:
            print('[Current] Insufficient funds or invalid amount.')


# What happens if a subclass forgets to implement a method?
class BrokenAccount(Account):
    pass   # forgot to implement deposit() and withdraw()

broken = BrokenAccount('Test', 999, 0)
try:
    broken.deposit(100)   # will raise NotImplementedError
except NotImplementedError as e:
    print(f'Error: {e}')

print()
# Correct usage
savings = SavingsAccount('Alice', 'SA123', 100)
current = CurrentAccount('Bob', 'CA456', 200)

savings.deposit(50)
savings.withdraw(20)
current.deposit(100)
current.withdraw(50)

Error: Subclasses must implement their own deposit() method

[Savings] Deposited $50. Balance: $150
[Savings] Withdrew $20 + $1 fee. Balance: $129
[Current] Deposited $100. Balance: $300
[Current] Withdrew $50. Balance: $250


### Big Tech Example: Payment Gateway as a Plugin Architecture

This is one of the most important uses of abstraction in real production software.

Companies like Shopify, Airbnb, and Amazon need to support many payment providers(Stripe, PayPal, Square, Adyen) without rewriting their checkout logic each time a new provider is added.

The solution: define a `PaymentGateway` **base class** with the required methods. Any new payment provider just implements those methods and the rest of the system works automatically.

This pattern is called the **Open/Closed Principle**:
- **Open** for extension : add a new gateway without touching existing code
- **Closed** for modification : `OrderProcessor` never needs to change

In [279]:
class PaymentGateway:
    # Base class: defines the contract that every gateway must follow
    # Any class that inherits from this must implement the three following methods

    def charge(self, amount, token):
        raise NotImplementedError('Every payment gateway must implement charge()')

    def refund(self, transaction_id, amount):
        raise NotImplementedError('Every payment gateway must implement refund()')

    def get_transaction(self, transaction_id):
        raise NotImplementedError('Every payment gateway must implement get_transaction()')

    def charge_with_tax(self, amount, token, tax_rate=0.1):
        # This method is implemented in the base class as it's a shared logic
        # It calls self.charge() which each gateway overrides
        total = round(amount * (1 + tax_rate), 2)
        print(f'Charging ${total} (${amount} + {int(tax_rate * 100)}% tax)')
        return self.charge(total, token)


class StripeGateway(PaymentGateway):
    def charge(self, amount, token):
        txn_id = 'ch_stripe_' + str(abs(hash(token)) % 10000) # hash(token) takes the token passed by us and creates a 4-digit number from it
        return {'status': 'success', 'id': txn_id, 'amount': amount, 'gateway': 'Stripe'}

    def refund(self, transaction_id, amount):
        print(f'[Stripe] Refunding ${amount} for transaction {transaction_id}')
        return True  # this is done to indicate that refund was done successfully or not, so that the calling function can act accordingly

    def get_transaction(self, transaction_id):
        return {'id': transaction_id, 'gateway': 'Stripe', 'status': 'settled'}


class PayPalGateway(PaymentGateway):
    def charge(self, amount, token):
        txn_id = 'PAY-' + str(abs(hash(token)) % 10000)  # hash(token) takes the token passed by us and creates a 4-digit number from it
        return {'status': 'success', 'id': txn_id, 'amount': amount, 'gateway': 'PayPal'}

    def refund(self, transaction_id, amount):
        print(f'[PayPal] Refunding ${amount} for transaction {transaction_id}')
        return True

    def get_transaction(self, transaction_id):
        return {'id': transaction_id, 'gateway': 'PayPal', 'status': 'completed'}


class OrderProcessor:
    def __init__(self, gateway):
        self.gateway = gateway  # stores that gateway object inside the instance so process_order() can use it

    def process_order(self, order_amount, card_token, customer_name):
        print(f'\nProcessing order for {customer_name}...')
        result = self.gateway.charge_with_tax(order_amount, card_token)
        if result['status'] == 'success':
            print(f"Order complete! [{result['gateway']}] Transaction ID: {result['id']}")
        return result

In [281]:
# Shopify uses Stripe for one merchant, PayPal for another — same OrderProcessor
stripe_processor = OrderProcessor(StripeGateway())
stripe_processor.process_order(89.99, 'tok_visa_4242', 'Alice') #here we are taking a token as an example ourselves, but in real system  
                                                                #Stripe will store your card details and generate token realted to it

paypal_processor = OrderProcessor(PayPalGateway())
paypal_processor.process_order(89.99, 'tok_pp_sandbox', 'Bob')  #here also we are taking a token as an example


Processing order for Alice...
Charging $98.99 ($89.99 + 10% tax)
Order complete! [Stripe] Transaction ID: ch_stripe_6040

Processing order for Bob...
Charging $98.99 ($89.99 + 10% tax)
Order complete! [PayPal] Transaction ID: PAY-3914


{'status': 'success', 'id': 'PAY-3914', 'amount': 98.99, 'gateway': 'PayPal'}

#### What Is a Token, that is being used?

A token is a randomly generated string that acts as a safe substitute for a customer's actual credit card number. Instead of passing around the real card number (4111 1111 1111 1234), the payment gateway replaces it with a meaningless string like tok_visa_4242 and that's what flows through your entire system.

This process is called tokenization.

---
## Summary

| Concept | What it does | Real-world analogy |
|---------|-------------|-------------------|
| **Classes & Objects** | Blueprint + instances created from it | House blueprint → actual houses |
| **`__init__`** | Sets up the object when it is created | Filling out a form when opening a bank account |
| **`self`** | The object referring to itself | "I" in a sentence |
| **`__str__`** | Controls what `print()` shows | The label on a product |
| **Class Variables** | Shared across ALL objects | Interest rate applied to every account |
| **Instance Variables** | Unique per object | Your personal bank balance |
| **Encapsulation** | Bundle data + control access | ATM machine where you press buttons, and not touch the wires |
| **Getter / Setter** | Controlled access to private data | Bank teller : you don't reach into the vault yourself |
| **Inheritance** | Reuse + extend a parent class | A SavingsAccount is also an Account |
| **`super()`** | Call the parent class method | Passing work up the chain to your manager |
| **Polymorphism** | Same method name, different behavior | `.speak()` → Cat says Meow, Dog says Woof |
| **Duck Typing** | If it has the method, use it | Any payment gateway with `.process()` works |
| **Overloading** | Same method name, flexible parameters | `add(5)`, `add(5, 10)`, `add(5, 10, 15)` all work |
| **Operator Overloading** | Redefine `+`, `==`, `<` for your class | `acc1 + acc2` creates a joint account |
| **Abstraction** | Define WHAT, leave HOW to subclasses | like Brake pedal where you don't need to know about hydraulics |
| **NotImplementedError** | Enforce that subclasses implement a method | A contract that says "you must fill this in" |